python/src/equistore/operations/reduce.py

```diff
@@ -0,0 +1,99 @@
+    """
+    for isample in sample_name:
+        if isample not in block.samples.names:
+            raise ValueError(
+                "the values in sample_names \
+                             should be contained in block1.samples.names"
+            )
+
+    # get the index of the selected sample
+    sample_selected = [block.samples.names.index(isample) for isample in sample_name]
+    # create a dictionary with the lists of all the index with the same sample_selected
+    samples = block.samples.view(dtype=np.int32).reshape(block.samples.shape[0], -1)
+    s, index = np.unique(samples[:, sample_selected], return_inverse=True, axis=0)
+    values_result = _dispatch.index_add(block.values, s.shape[0], index)
+
+    # if s.size == 1:
+    #    s = s.reshape(1, len(s))
+
+    result_block = TensorBlock(
+        values=values_result,
+        samples=Labels(
+            [i for i in sample_name], np.array([i for i in s], dtype=np.int32)
+        ),  # TODO guardare
+        components=block.components,
+        properties=block.properties,
+    )
+
+    for parameter, gradient in block.gradients():
+        # here I copy because I want to modify the samples array
+        samples = (
+            gradient.samples.view(dtype=np.int32)
+            .reshape(gradient.samples.shape[0], -1)
+            .copy()
+        )
+        for i in range(len(samples)):
+            samples[i, 0] = index[samples[i, 0]]
+
+        s_gradient, index_gradient = np.unique(
+            samples[:, :], return_inverse=True, axis=0
+        )
+        data_result = _dispatch.index_add(
+            gradient.data, s_gradient.shape[0], index_gradient
+        )
+
+        result_block.add_gradient(
+            parameter,
+            data_result,
+            Labels(
+                ["sample"] + [i for i in gradient.samples.names[1:]],
+                np.array([i for i in s_gradient], dtype=np.int32),
+            ),
+            gradient.components,
+        )
+
+    return result_block
+
+
+def reduce(tensor: TensorMap, sample_name: list) -> TensorMap:
+    """Create a new :py:class:`TensorMap` with the same keys as
+    as the input `tensor`, and each :py:class:`TensorBlock`
+    is obtained summing the corresponding input :py:class:`TensorBlock`
+    over the rows with the same sample_names.
+
+    For example if sample_name = ["structure"] it sum over
+    all the sample with the same structure
+    if sample_name = ["structure", "center"] it sums over all the rows with
+    the same values for the "structure" and the "center"
+
+    :param tensor:  input :py:class:`TensorMap`
+    :param sample_name: names of samples to sum
+    """
+
+    try:
+        blocks = []
+        for key, block in tensor:
+            blocks.append(_reduction_sample_block(block=block, sample_name=sample_name))
+        return TensorMap(tensor.keys, blocks)
+    except ValueError:
+        print("it is not possible to perform the reduction for all the blocks")
```

Comment:
Printing when you caught an exception is not a good practise: the user will likely miss the error and can not do anything about it. Why not just let the exception go all the way to the user here?